### PR TITLE
Fix using dashes in AWS resource tags, add unit tests

### PR DESCRIPTION
**Pull request by Jalle19**
This regular expression has been broken before, it's time it gets tested properly

Closes: #8764

I'm not sure whether `naming.js` is the best location to move the regular expression to, but it was the best I could find given that I'm not familiar with the codebase.

### CODE REVIEW

**Comment by Jalle19** [comment]
Added tests for the 128 character maximum length too

**Comment by Jalle19** [comment]
@pgrzesik escaping the dash did the trick indeed. The dot needed escaping too (I'm actually not 100% sure about that, but it might be safest to escape it anyway to avoid confusion).

File: lib/plugins/aws/lib/naming.js [hunk 1]
```diff
@@ -638,4 +638,9 @@ module.exports = {
   getHttpApiLogGroupName() {
     return `/aws/http-api/${this.getStackName()}`;
   },
+  // Tags
+  getResourceTagNameRegex() {
```
Comment (pgrzesik) [inline lib/plugins/aws/lib/naming.js:642]: `naming.js` is probably not a good place as the intention behind this module is to be used when creating names for CloudFormation resources, I think we should go with separate module for such regexes - @medikoo, what do you think about this approach in general?

**Comment by pgrzesik** [comment]
Thanks @Jalle19 for submitting this PR :bow: I'd like to wait for @medikoo's response before moving forward as I'm not sure yet about the approach

File: lib/plugins/aws/lib/naming.js (hunk 1 above)
Comment (Jalle19) [reply lib/plugins/aws/lib/naming.js:642]: Yeah - I looked through the code for other files where regular expressions where defined, but the only one I found that was actually defined in any sense (i.e. not just defined directly where it was used) was something about ALB naming.

A `regexp.js` on the same level as `naming.js` could do the trick. I'm sure there are more regular expressions that could use some basic tests.

Comment (medikoo) [reply lib/plugins/aws/lib/naming.js:642]: @Jalle19 great thanks for an attempt on solving this.

The way we handle schema configuration, is that regex should simply be put into schema (let's not mix into `naming`, as it serves very different purpose).

Concerning testing, we do not practice testing schema validation directly.

We test it indirectly by simply testing that defined properties produce desired effect.

In this case it'll be great to update (or add) a test which tests the property that relies on definition (e.g. `provider.stackTags`) and confirm that it accepts the tag with format that was previously rejected

**Review by medikoo — CHANGES_REQUESTED** [review]

File: lib/plugins/aws/lib/naming.js (hunk 1 above)
Comment (Jalle19) [reply lib/plugins/aws/lib/naming.js:642]: @medikoo so I should not introduce a new `regexp.js` file then?

**Comment by Jalle19** [comment]
I found https://github.com/serverless/serverless/pull/8051, but I can't dig up the tests that were inevitably scrapped from the pull request. Maybe reconsider not testing these validations at all? Clearly they are the source of multiple bugs, and complicating testing makes the barrier to entry for fixing them higher. I mean, I've already fixed the bug and written tests.

I agree that simple validations don't need testing because the logic is part of `ajv`, but regular expressions are part of this code and should be tested here.

File: lib/plugins/aws/lib/naming.js (hunk 1 above)
Comment (medikoo) [reply lib/plugins/aws/lib/naming.js:642]: @Jall19 as this regex is used just in this single place, let's just put it inline there.

We do not practice secluding regular expression into some other dedicated file.

Comment (Jalle19) [reply lib/plugins/aws/lib/naming.js:642]: There are quite a few complicated regular expressions used in the rule definitions that could use some rudimentary testing, especially for edge cases.

**Comment by medikoo** [comment]
> Maybe reconsider not testing these validations at all? 

We don't want to fall into practice of writing tests for regular expressions confirming that they support all intended notations.
Value of such tests is controversial and they rise maintenance cost of already big project.

Approach we have is that we test properties of which format is validated through regular expression.

Here the real problem is not that regex was invalid, but that we didn't support given notation for a property, and that should simply addressed by (1) improving the regex which backs validation of that property, and (2) ensuring a regression test, that confirms, that given notation is now supported for given property.

So when testing is concerned I would simply add previously not supported notation e.g. in this test: https://github.com/serverless/serverless/blob/420e93740f1e9bffc285559b2567379f550f28af/test/unit/lib/plugins/aws/package/compile/functions.test.js#L379-L414 we may also add a comment there that it tests regression reported at <github link> (so at some cleanup by accident it's not removed)

**Comment by Jalle19** [comment]
@medikoo I changed the way the test is done, please review again

File: test/unit/lib/plugins/aws/package/compile/functions.test.js [hunk 2]
```diff
@@ -2244,6 +2244,39 @@ describe('lib/plugins/aws/package/compile/functions/index.test.js', () => {
...
+              provider: {
+                stackTags: {
+                  [tagKey]: 'foo',
+                },
+              },
+            },
+            cliArgs: ['package'],
+          });
+
+          expect.fail('an error should have been thrown');
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/package/compile/functions.test.js:2272]: Let's not confirm on specific schema validation errors in tests (otherwise we'll end with 10k+ scenarios to confirm). We have a sanity check that validation engine is configured properly, and schema engine on its own is tested extensively. I believe that's good enough.

What we should test here is simply that configuring tag as `foo:bar-baz` works as expected, and as I indicated in other comment, we already test function tagging here: https://github.com/serverless/serverless/blob/420e93740f1e9bffc285559b2567379f550f28af/test/unit/lib/plugins/aws/package/compile/functions.test.js#L379-L414

Let's simply add there such tag, and confirm it got configured as expected in CF template

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @Jalle19 ! It's way better, I've just proposed a different approach to testing, which we practice in context of this project.

File: test/unit/lib/plugins/aws/package/compile/functions.test.js (hunk 2 above)
Comment (Jalle19) [reply test/unit/lib/plugins/aws/package/compile/functions.test.js:2272]: Adding a new tag here accomplishes nothing - the test passes regardless of whether the regular expression is fixed or not.

Comment (medikoo) [reply test/unit/lib/plugins/aws/package/compile/functions.test.js:2272]: Sorry, I didn't look carefully, and pasted you link to wrong test (one to be discarded).

Just add tag here: https://github.com/serverless/serverless/blob/7d2ca04d7f72981c422fb8f9c25ae4699aedffba/test/unit/lib/plugins/aws/package/compile/functions.test.js#L1466-L1469

Fact that it works will be automatically confirmed here: https://github.com/serverless/serverless/blob/7d2ca04d7f72981c422fb8f9c25ae4699aedffba/test/unit/lib/plugins/aws/package/compile/functions.test.js#L1553-L1562

**Comment by Jalle19** [comment]
@medikoo I removed the tests and added just a few more defined tags to the existing test case. I'm not a fan of this solution since it can only test that valid values are allowed, not that invalid values are disallowed.

File: lib/plugins/aws/provider.js [hunk 3]
```diff
@@ -521,7 +521,7 @@ class AwsProvider {
           awsResourceTags: {
             type: 'object',
             patternProperties: {
-              '^(?!aws:)[\\w./=+:-_\\x20]{1,128}$': {
+              '^(?!aws:)[\\w\\./=+:\\-_\\x20]{1,128}$': {
```
Comment (medikoo) [inline lib/plugins/aws/provider.js:524]: There's no need to escape `.`

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @Jalle19 ! It looks great now, I've just proposed to reduce noise in regex and clear obsolete escape

File: lib/plugins/aws/provider.js (hunk 3 above)
Comment (Jalle19) [reply lib/plugins/aws/provider.js:524]: I figured it would be clearer to escape it, but okay, I'll revert

**Comment by Jalle19** [comment]
This should be good to go now IMO, can't make the change any smaller

**Review by medikoo — APPROVED** [review]
Thank you @Jalle19, looks great!
